(feat) Computer feature: use library data model if selected path is watched by library

ronso0 opened the pull request:
Fixes #14683 
Related: #11026 #9897

### TL:DR
If the selected path is watched by library, we show it like the other track views (Tracks, Crates, ..).
This gives us the same the view (colums) and editing capabilities.
Watched paths are currently marked with an ~~orange~~ **cyan dot**:
![image](https://github.com/user-attachments/assets/6d039b59-746f-451e-ac42-f3a9956950c9)


This is really handy if you want to browse your library by path!
Supposed to enhance the **Sidebar Bookmarks** use case #14508.

### Details
If the path is watched, we use an alternative model (derived from `LibraryTableModel`, uses track data).
Luckily, LibraryTableModel already allows to use an extra search filter (not visible in the searchbox) which allows to filter tracks by directory. (such a filter is also used in _Analyze_ view)
Since BrowseFeature tells the library to load its model on click, switching between the file / track model works flawlessly.

This also introduces `SidebarItemDelegate` to do the painting.

**Note**:
The header is different for the file / track views (eg. the file view does not have the rating column), but the views can be made somewhat similar by adjusting the headers once.

- [x] adopt search query when switching between file / track model
- [x] add subtle decoration if path item is watched (probably with sidebar delegate from #14508)
- [ ] cache `isWatchedPath` property to avoid string comparison when expanding the tree?
- [ ] don't rebuild child tree if watched path is expanded? (update on demand with "Refresh directory tree")
try to adopt solution from #14508
- [x] update TreeItems' `isWatchedPath` flag when directories have been added to or removed from library
- [ ] also switch view model in that case, if current view is affected?

### TODO
- [ ] ~~use ibrary view in Recording, too~~ postponed, will happen as followup
- [x] verify all commits build
- [x] test on Linux
- [ ] test on Windows
- [ ] test on macOS
- [ ] test with watched network shares on all OS
- [ ] make it work with #14479, ie.

ronso0 commented:
Hehe, I tested this in a session with my personal branch (ie. on top of Sidebar Bookmakrs PR #14508) and it all works as expected!

ronso0 committed:
Browse feature: remove TODO, add comment

ronso0 committed:
Browse feature: add library view comment to rescan dialog

ronso0 commented:
Cool, I merged this and #14508 **Sidebar Bookmarks** into my performace branch, and it works nicely!
Just had to adjust the columns, then I had the same tracks view as in Tracks.

Eve00000 commented:
Testing on Windows
- Quick Links & Devices only give a blue dot
- No data from dbase shown
see image, properties window is opened from deck 2 genre & comment in dbase, not in trackview
![afbeelding](https://github.com/user-attachments/assets/d1af8925-90b2-48ca-8e36-8e34795ca107)

ronso0 commented:
Can you reproduce the crash on pure main?

FYI the entire 'library view' thing with all the features is actually just a filter for track file directory. So, if -for some reason- a non-library directory is activated with the library view, it won't show any tracks (except maybe single tracks from that directory that have been loaded earlier).

Eve00000 commented:
Strange ... 
What I did on Windows:
- clean build folder  -> build main -> no issues
- clean build folder -< build PR -> 10 restarts of Mixxx browsing through quick links & devices 
-> no crash .... I can't reproduce.

Other remark: no files displayed for folders with special signs like
F:\Music\#9999 - DJ\#0500 - Artists\'t Hof Van Commerce •\'t Hof Van Commerce - En In Izzegem . . _
(has 2 files)
while all files are displayed in the folder 
F:\Music\#9999 - DJ\#0500 - Artists\'t Hof Van Commerce •\'t Hof Van Commerce - Rocky 7

ronso0 commented:
> Other remark: no files displayed for folders with special signs like
> F:\Music#9999 - DJ#0500 - Artists't Hof Van Commerce •'t Hof Van Commerce - En In Izzegem . . _

What do you see when you go to Tracks and enter that in the searchbar?
"F:\Music#9999 - DJ#0500 - Artists't Hof Van Commerce •'t Hof Van Commerce - En In Izzegem"
that or
location:"F:\Music#9999 - DJ#0500 - Artists't Hof Van Commerce •'t Hof Van Commerce - En In Izzegem"

ronso0 commented:
> Looks like you fixed it, I only have blue dots, is that correct?

Yup, that's how it's supposed to look like. 
Dots for all children of a watched path are redundant (as we don't have 'exclude paths' yet, I mean that's possible but it would check each path's children on parent expandation) but it makes the situation clear if you're deep down in the trees : )

Btw, I think we need some sort of clarifying comment in the Computer root view, see latest Digital DJ Tips review.

Eve00000 commented:
Searching for the location in searchlinedit works for
F:/Music/#9999 - DJ/#0500 - Artists/'t Hof Van Commerce •/'t Hof Van Commerce - En In Izzegem . . _
location:"F:/Music/#9999 - DJ/#0500 - Artists/'t Hof Van Commerce •/'t Hof Van Commerce - En In Izzegem . . _"
location:"F:/Music/#9999 - DJ/#0500 - Artists/'t Hof Van Commerce •/'t Hof Van Commerce - En In Izzegem"
![afbeelding](https://github.com/user-attachments/assets/5b1e87cb-f1eb-45bf-8d30-86d8507a80f8)

view on the track_locations

Eve00000 commented:
> Yup, that's how it's supposed to look like.

ah, I thoufgt the had to be orange, no idea were I got that.

> Btw, I think we need some sort of clarifying comment in the Computer root view, see latest Digital DJ Tips review.

Correct, I'm making a list of issues that came up in reviews, I'll add them to the issues when ready.

ronso0 commented:
> Searching for the location in searchlinedit works for
F:/Music/#9999 - DJ/#0500 - Artists/'t Hof Van Commerce •/'t Hof Van Commerce - En In Izzegem . . _

Please post the full directory name. I'll try to reproduce.

ronso0 commented:
Okay, so the issue is the traling `_`
I have no clue why..
will keep digging.

ronso0 commented:
I fixed the `_` issue. It occured because I used SQL `LIKE` which interprets trailing _ as wildcard, which for some reason didn't match `_`.

____
The false-positive "has children" is a different issue, and no regression of this PR.
That happens because, on Linux and macOS, consider a (sym)link in a directory a potential directory, but don't actually check if it points to a directory. I'll propose a fix in a separate PR.

ronso0 committed:
debug/trace

ronso0 commented:
Thanks for the hints, but No I did not look into the library-redesign branch.
And I did not think of any kind of grouping feature.
The goal was simply getting the track view (not file view) into the tree, ie. allow enhanced **tree browsing**: file view for unwatched folders, track view for watched.
And not complicate the library too much.

Grouping is unrelated IMO, though a nice feature IIUC.
There was also the idea to list the watched folders in **Tracks**.

The recursive option is something I planned to look into as well, though not as general option but per folder on demand.

Actually, the current solution was pretty easy to implement.
Just the extra symlink feature was pretty tedious, but is also working smoothly and reliably now.

One thing I forgot I think is updating the current folder (library) view after rescan. Will do this.

ronso0 commented:
> We have learned that the expectation is to see the file information from the database and have them editable. This is not only an issue of watched folders.
> 
> Maybe we can combine it and hide some details and hopefully simplify your code:
> 
> How about this:
> 
> 1. Browse the folder as before.
> 2. Look up each location in the database first
> 3. If exists Use the database metadata.
> 4. If not exists, parse the file metadata without adding the file to the database.
> 5. When a track is played, it is added to the database
> 6. When metadata is edited the track is added to the database
> 
> For the rare "seeing the truth" use case, we may do a global preferences option "show file metadata" which behaves like before.

Well, from what I learned from digging through the Browse model / Browse thread that's not gonna be easy.

My main uses case are primarily:
1. attach guest DJ's USB drive, add it to the library, scan it -> access it via its tree
2. browse my own library directories when file (location) don't work or are to tedious to create

ronso0 commented:
> > The recursive option is something I planned to look into as well, though not as general option but per folder on demand.
> 
> Can you explain this?

With this directory
```
Wordsound
├─ track01
├─ track02
├─ track03
├─ track04
├─ Album1
│  ├─ track01
│  ├─ track02
│  ├─ track03
│  ├─ track04
│  └─ track05
└─ below the radar
   ├─ ...
```
the directory filter would currently only show top-level tracks of Wordsound.
With the "recursive" option it would show all tracks, incl. Album1, below the radar an other

ronso0 commented:
> Is there a use case to add watched folders via the browse feature and indicate that a certain folder is watched? For my feeling the majority of users does not care about it and just add one or two root folders in the library preferences.

This is not a new feature.
I find it useful for temp devices like USB drives. (save going to preferences)
Just add and press Ctrl+Shift+L to rescan.

ronso0 commented:
Yes, that's it.
Root cause is that we `extraFilter` is [wrapped into a SqlNode](https://github.com/mixxxdj/mixxx/blob/086f48e8a11e4c35748bced4ab39dfa799d55402/src/library/searchqueryparser.cpp#L351) and [`SqlNode::match()` always returns true](https://github.com/mixxxdj/mixxx/blob/086f48e8a11e4c35748bced4ab39dfa799d55402/src/library/searchquery.h#L263-L267).
Hence the check
```
bool shouldBeInResultSet = searchQuery.isEmpty() ||
        pQuery->match(pTrack);
```
that is supposed to add dirty tracks is pointless / counter-productive.

I'll file a separate issue if I don't find a quick fix myself.

ronso0 commented:
Great :tada: 

This is a massive change set, so I understand that's it's been sitting so long.
I skimmed through the commits again and noticed I still need t clean up a few things and can improve comments.
I'll try to do that soonish and let you know when I'm done (remove Draft state).

Let me know if you wanna start reviewing before I'm done.
(I made numerous fixups in my production branch, and they usually create complex conflicts when rebasing -- that's why I'd prefer to fixup during rebase and force-push)

ronso0 committed:
Browse feature: use parented_ptr, forward-declare, rm unused includes

ronso0 committed:
Browse feature: use Library model when path is watched (in library)

library directory view currently uses recursive mode

ronso0 committed:
Browse feature: use symlink map and update it and `watched` flags when library directories changed

accomplished with QFuture/QFutureWatcher as well as QPromise to avoid trouble with potential multiple calls

ronso0 committed:
Browse feature: decorate watched path items

ronso0 committed:
Browse feature: show root view when clicking Quick Links or Removable Devices

ronso0 commented:
I rebased and  polished / cleaned up commits. All build and work as expected.
IMO this is now ready for review.

ronso0 commented:
Been using this for quite some time now, since two weeks also on a friend's machine, and it turns out it would be cool to have a shortcut for rescanning a specific directory (something like #15940).

ronso0 commented:
> I rebased and polished / cleaned up commits. All build and work as expected.
IMO this is now ready for review.

Did some more clean up, and, more importantly ironed out some issues with the sidebar model as well as (potential) issues with the QFutureWatcher used for the symlink map (which was the only part where I still had bad feelings..).
Check it out if you dare : )
